Version file is not generated since version 8 on CI without any error

Since `setuptools_scm` release 8 I have observed a test crash because the version file is not generated. 

```
   /tmp/tox/py311-PyQt5-all/lib/python3.11/site-packages/PartSeg/__init__.py:4: in <module>
      from PartSeg.version import version
  E   ModuleNotFoundError: No module named 'PartSeg.version'
``` 

When it is configured in pyproject.toml: https://github.com/4DNucleome/PartSeg/blob/bcaf0aad2e80edbf25e42859f8a0722074f0c68b/pyproject.toml#L1-L7

But when I pin `setuptools_scm` below 8 all problems disappear.

Here is log with `SETUPTOOLS_SCM_DEBUG` set to 1. 
https://github.com/4DNucleome/PartSeg/actions/runs/6273817081/job/17038040421?pr=1020

I cannot reproduce it locally. I do not have any idea how to debug this.

Hello 👋 

I am also getting a similar error, both in version 8 and in previous versions (tested on 7.1.0 and 8.0.3). It only started happening in the last few days, and it coincided with a new release of an Ubuntu 22.04 image (20230917.1.0, could be coincidental/unrelated) on GitHub Actions.

Example failed build (7.1.0): https://github.com/mozilla/mozregression/actions/runs/6240525450/job/17049062792.
Link to pyproject.toml: https://github.com/mozilla/mozregression/blob/main/pyproject.toml#L6

```
mozregression/__init__.py:1: in <module>
    from .version import version
E   ModuleNotFoundError: No module named 'mozregression.version'
```

Edit: Seems to affect most versions of Python (tested 3.7 to 3.11), and only on Ubuntu 22.04 and Windows for some reason. macOS and Ubuntu 20.04 appear to not be affected by this issue.

> @zzzeid what setuptools version is made available

Seems like they all are using 68.0.0. I also checked the old build from last week that was successful, uses the same version.

[Successful build](https://github.com/mozilla/mozregression/actions/runs/6277839473/job/17051249935) (Python 3.9, Ubuntu 20.04): setuptools==68.0.0
[Failed build](https://github.com/mozilla/mozregression/actions/runs/6277839473/job/17051251457) (Python 3.10, Ubuntu 22.04): setuptools==68.0.0
[Failed build](https://github.com/mozilla/mozregression/actions/runs/6277839473/job/17051249483) (Python 3.11, Windows): setuptools==68.0.0
[Successful build](https://github.com/mozilla/mozregression/actions/runs/6277839473/job/17051248660) (Python 3.11, macOS): setuptools==68.0.0

> can you try what happens if setup_requires is removed?

I will try that and let you know.

Edit: [same](https://github.com/mozilla/mozregression/actions/runs/6278635500/job/17052810441?pr=1470) error after removing `setup_requires`.

(sorry, realized that you can't see the detail on those runs, but the error is exactly the same, pasted below).
```
============================= test session starts ==============================
platform linux -- Python 3.8.18, pytest-7.4.2, pluggy-1.2.0
PySide6 6.5.1.1 -- Qt runtime 6.5.1 -- Qt compiled 6.5.1
rootdir: /home/runner/work/mozregression/mozregression
plugins: mock-3.11.1, qt-4.2.0
collected 0 items / 1 error

==================================== ERRORS ====================================
________________________ ERROR collecting test session _________________________
env/lib/python3.8/site-packages/_pytest/config/__init__.py:641: in _importconftest
    mod = import_path(conftestpath, mode=importmode, root=rootpath)
env/lib/python3.8/site-packages/_pytest/pathlib.py:567: in import_path
    importlib.import_module(module_name)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1014: in _gcd_import
    ???
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
env/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:178: in exec_module
    exec(co, module.__dict__)
tests/unit/conftest.py:3: in <module>
    from mozregression import build_range
mozregression/__init__.py:1: in <module>
    from mozregression.version import version
E   ModuleNotFoundError: No module named 'mozregression.version'
=========================== short test summary info ============================
ERROR  - ModuleNotFoundError: No module named 'mozregression.version'
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.57s ===============================
Error: Process completed with exit code 2.
```

--- a/changelog.d/20231002_161552_opensource_fix_938_self_reinit_fixup.md
+++ b/changelog.d/20231002_161552_opensource_fix_938_self_reinit_fixup.md
@@ -1,0 +1,6 @@
+
+### Changed
+
+- ensure the setuptools version keyword correctly load pyproject.toml configuration
+- add build and wheel to the test requirements for regression testing
+- move internal toml handling to own module

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -57,6 +57,7 @@ rich = [
   "rich",
 ]
 test = [
+  "build",
   "pytest",
   "rich",
 ]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -60,6 +60,7 @@ test = [
   "build",
   "pytest",
   "rich",
+  "wheel",
 ]
 toml = [
 ]

--- a/src/setuptools_scm/_config.py
+++ b/src/setuptools_scm/_config.py
@@ -125,11 +125,14 @@ class Configuration:
         not contain the [tool.setuptools_scm] section.
         """
 
-        pyproject_data = _read_pyproject(name, _load_toml=_load_toml)
+        pyproject_data = _read_pyproject(
+            name, _load_toml=_load_toml, require_section=_require_section
+        )
         args = _get_args_for_pyproject(pyproject_data, dist_name, kwargs)
 
         args.update(read_toml_overrides(args["dist_name"]))
-        return cls.from_data(relative_to=name, data=args)
+        relative_to = args.pop("relative_to", name)
+        return cls.from_data(relative_to=relative_to, data=args)
 
     @classmethod
     def from_data(

--- a/src/setuptools_scm/_config.py
+++ b/src/setuptools_scm/_config.py
@@ -5,8 +5,8 @@ import dataclasses
 import os
 import re
 import warnings
+from pathlib import Path
 from typing import Any
-from typing import Callable
 from typing import Pattern
 from typing import Protocol
 
@@ -114,7 +114,6 @@ class Configuration:
         cls,
         name: str | os.PathLike[str] = "pyproject.toml",
         dist_name: str | None = None,
-        _load_toml: Callable[[str], dict[str, Any]] | None = None,
         _require_section: bool = True,
         **kwargs: Any,
     ) -> Configuration:
@@ -125,9 +124,7 @@ class Configuration:
         not contain the [tool.setuptools_scm] section.
         """
 
-        pyproject_data = _read_pyproject(
-            name, _load_toml=_load_toml, require_section=_require_section
-        )
+        pyproject_data = _read_pyproject(Path(name), require_section=_require_section)
         args = _get_args_for_pyproject(pyproject_data, dist_name, kwargs)
 
         args.update(read_toml_overrides(args["dist_name"]))

--- a/src/setuptools_scm/_config.py
+++ b/src/setuptools_scm/_config.py
@@ -115,6 +115,7 @@ class Configuration:
         name: str | os.PathLike[str] = "pyproject.toml",
         dist_name: str | None = None,
         _load_toml: Callable[[str], dict[str, Any]] | None = None,
+        _require_section: bool = True,
         **kwargs: Any,
     ) -> Configuration:
         """

--- a/src/setuptools_scm/_entrypoints.py
+++ b/src/setuptools_scm/_entrypoints.py
@@ -45,7 +45,7 @@ log = _log.log.getChild("entrypoints")
 
 
 def version_from_entrypoint(
-    config: Configuration, entrypoint: str, root: _t.PathT
+    config: Configuration, *, entrypoint: str, root: _t.PathT
 ) -> version.ScmVersion | None:
     from .discover import iter_matching_entrypoints
 

--- a/src/setuptools_scm/_get_version_impl.py
+++ b/src/setuptools_scm/_get_version_impl.py
@@ -32,18 +32,22 @@ def parse_scm_version(config: Configuration) -> ScmVersion | None:
                 )
             return parse_result
         else:
-            entrypoint = "setuptools_scm.parse_scm"
-            root = config.absolute_root
-            return _entrypoints.version_from_entrypoint(config, entrypoint, root)
+            return _entrypoints.version_from_entrypoint(
+                config,
+                entrypoint="setuptools_scm.parse_scm",
+                root=config.absolute_root,
+            )
     except _run_cmd.CommandNotFoundError as e:
         _log.exception("command %s not found while parsing the scm, using fallbacks", e)
         return None
 
 
 def parse_fallback_version(config: Configuration) -> ScmVersion | None:
-    entrypoint = "setuptools_scm.parse_scm_fallback"
-    root = config.fallback_root
-    return _entrypoints.version_from_entrypoint(config, entrypoint, root)
+    return _entrypoints.version_from_entrypoint(
+        config,
+        entrypoint="setuptools_scm.parse_scm_fallback",
+        root=config.fallback_root,
+    )
 
 
 def parse_version(config: Configuration) -> ScmVersion | None:

--- a/src/setuptools_scm/_integration/pyproject_reading.py
+++ b/src/setuptools_scm/_integration/pyproject_reading.py
@@ -46,21 +46,29 @@ def read_pyproject(
     name: str | os.PathLike[str] = "pyproject.toml",
     tool_name: str = "setuptools_scm",
     _load_toml: TOML_LOADER | None = None,
-    require_secion: bool = True,
+    require_section: bool = True,
 ) -> PyProjectData:
     if _load_toml is None:
         _load_toml = lazy_toml_load
-    with open(name, encoding="UTF-8") as strm:
-        data = strm.read()
-    defn = _load_toml(data)
+    try:
+        with open(name, encoding="UTF-8") as strm:
+            data = strm.read()
+    except FileNotFoundError:
+        if require_section:
+            raise
+        else:
+            log.debug("%s missing, presuming emptry as section is not required", name)
+            defn = {}
+    else:
+        defn = _load_toml(data)
     try:
         section = defn.get("tool", {})[tool_name]
     except LookupError as e:
         error = f"{name} does not contain a tool.{tool_name} section"
-        if require_secion:
+        if require_section:
             raise LookupError(error) from e
         else:
-            log.warn(error)
+            log.warning(error)
             section = {}
 
     project = defn.get("project", {})

--- a/src/setuptools_scm/_integration/pyproject_reading.py
+++ b/src/setuptools_scm/_integration/pyproject_reading.py
@@ -9,10 +9,13 @@ from typing import Dict
 from typing import NamedTuple
 from typing import TYPE_CHECKING
 
+from .. import _log
 from .setuptools import read_dist_name_from_setup_cfg
 
 if TYPE_CHECKING:
     from typing_extensions import TypeAlias
+
+log = _log.log.getChild("pyproject_reading")
 
 _ROOT = "root"
 TOML_RESULT: TypeAlias = Dict[str, Any]
@@ -43,6 +46,7 @@ def read_pyproject(
     name: str | os.PathLike[str] = "pyproject.toml",
     tool_name: str = "setuptools_scm",
     _load_toml: TOML_LOADER | None = None,
+    require_secion: bool = True,
 ) -> PyProjectData:
     if _load_toml is None:
         _load_toml = lazy_toml_load
@@ -52,7 +56,13 @@ def read_pyproject(
     try:
         section = defn.get("tool", {})[tool_name]
     except LookupError as e:
-        raise LookupError(f"{name} does not contain a tool.{tool_name} section") from e
+        error = f"{name} does not contain a tool.{tool_name} section"
+        if require_secion:
+            raise LookupError(error) from e
+        else:
+            log.warn(error)
+            section = {}
+
     project = defn.get("project", {})
     return PyProjectData(name, tool_name, project, section)
 

--- a/src/setuptools_scm/_integration/pyproject_reading.py
+++ b/src/setuptools_scm/_integration/pyproject_reading.py
@@ -1,29 +1,22 @@
 from __future__ import annotations
 
-import os
-import sys
 import warnings
-from typing import Any
-from typing import Callable
-from typing import Dict
+from pathlib import Path
 from typing import NamedTuple
-from typing import TYPE_CHECKING
 
 from .. import _log
 from .setuptools import read_dist_name_from_setup_cfg
+from .toml import read_toml_content
+from .toml import TOML_RESULT
 
-if TYPE_CHECKING:
-    from typing_extensions import TypeAlias
 
 log = _log.log.getChild("pyproject_reading")
 
 _ROOT = "root"
-TOML_RESULT: TypeAlias = Dict[str, Any]
-TOML_LOADER: TypeAlias = Callable[[str], TOML_RESULT]
 
 
 class PyProjectData(NamedTuple):
-    name: str | os.PathLike[str]
+    path: Path
     tool_name: str
     project: TOML_RESULT
     section: TOML_RESULT
@@ -33,46 +26,24 @@ class PyProjectData(NamedTuple):
         return self.project.get("name")
 
 
-def lazy_toml_load(data: str) -> TOML_RESULT:
-    if sys.version_info >= (3, 11):
-        from tomllib import loads
-    else:
-        from tomli import loads
-
-    return loads(data)
-
-
 def read_pyproject(
-    name: str | os.PathLike[str] = "pyproject.toml",
+    path: Path = Path("pyproject.toml"),
     tool_name: str = "setuptools_scm",
-    _load_toml: TOML_LOADER | None = None,
     require_section: bool = True,
 ) -> PyProjectData:
-    if _load_toml is None:
-        _load_toml = lazy_toml_load
-    try:
-        with open(name, encoding="UTF-8") as strm:
-            data = strm.read()
-    except FileNotFoundError:
-        if require_section:
-            raise
-        else:
-            log.debug("%s missing, presuming emptry as section is not required", name)
-            defn = {}
-    else:
-        defn = _load_toml(data)
+    defn = read_toml_content(path, None if require_section else {})
     try:
         section = defn.get("tool", {})[tool_name]
     except LookupError as e:
-        error = f"{name} does not contain a tool.{tool_name} section"
+        error = f"{path} does not contain a tool.{tool_name} section"
         if require_section:
             raise LookupError(error) from e
         else:
-            log.warning(error)
+            log.warning("toml section missing %r", error)
             section = {}
 
     project = defn.get("project", {})
-    return PyProjectData(name, tool_name, project, section)
+    return PyProjectData(path, tool_name, project, section)
 
 
 def get_args_for_pyproject(
@@ -86,7 +57,7 @@ def get_args_for_pyproject(
     if "relative_to" in section:
         relative = section.pop("relative_to")
         warnings.warn(
-            f"{pyproject.name}: at [tool.{pyproject.tool_name}]\n"
+            f"{pyproject.path}: at [tool.{pyproject.tool_name}]\n"
             f"ignoring value relative_to={relative!r}"
             " as its always relative to the config file"
         )
@@ -110,5 +81,5 @@ def get_args_for_pyproject(
                     f"root {section[_ROOT]} is overridden"
                     f" by the cli arg {kwargs[_ROOT]}"
                 )
-            section.pop("root", None)
+            section.pop(_ROOT, None)
     return {"dist_name": dist_name, **section, **kwargs}

--- a/src/setuptools_scm/_integration/setuptools.py
+++ b/src/setuptools_scm/_integration/setuptools.py
@@ -63,21 +63,27 @@ def _assign_version(
 _warn_on_old_setuptools()
 
 
+def _log_hookstart(hook: str, dist: setuptools.Distribution) -> None:
+    log.debug("%s %r", hook, vars(dist.metadata))
+
+
 def version_keyword(
     dist: setuptools.Distribution,
     keyword: str,
     value: bool | dict[str, Any] | Callable[[], dict[str, Any]],
 ) -> None:
-    if not value:
-        return
-    elif value is True:
+    if value is True:
         value = {}
     elif callable(value):
         value = value()
+
+    assert isinstance(value, dict), "version_keyword expects a dict or True"
     assert (
         "dist_name" not in value
     ), "dist_name may not be specified in the setup keyword "
     dist_name: str | None = dist.metadata.name
+    _log_hookstart("version_keyword", dist)
+
     if dist.metadata.version is not None:
         warnings.warn(f"version of {dist_name} already set")
         return
@@ -103,10 +109,7 @@ def version_keyword(
 
 
 def infer_version(dist: setuptools.Distribution) -> None:
-    log.debug(
-        "finalize hook %r",
-        vars(dist.metadata),
-    )
+    _log_hookstart("infer_version", dist)
     log.debug("dist %s %s", id(dist), id(dist.metadata))
     if dist.metadata.version is not None:
         return  # metadata already added by hook
@@ -120,6 +123,6 @@ def infer_version(dist: setuptools.Distribution) -> None:
     try:
         config = _config.Configuration.from_file(dist_name=dist_name)
     except LookupError as e:
-        log.exception(e)
+        log.warning(e)
     else:
         _assign_version(dist, config)

--- a/src/setuptools_scm/_integration/setuptools.py
+++ b/src/setuptools_scm/_integration/setuptools.py
@@ -9,7 +9,6 @@ from typing import Callable
 import setuptools
 
 from .. import _config
-from .._version_cls import _validate_version_cls
 
 log = logging.getLogger(__name__)
 
@@ -95,15 +94,9 @@ def version_keyword(
 
     if dist_name is None:
         dist_name = read_dist_name_from_setup_cfg()
-    version_cls = value.pop("version_cls", None)
-    normalize = value.pop("normalize", True)
-    tag_regex = _config._check_tag_regex(
-        value.pop("tag_regex", _config.DEFAULT_TAG_REGEX)
-    )
-    final_version = _validate_version_cls(version_cls, normalize)
 
-    config = _config.Configuration(
-        dist_name=dist_name, version_cls=final_version, tag_regex=tag_regex, **value
+    config = _config.Configuration.from_file(
+        dist_name=dist_name, _require_section=False, **value
     )
     _assign_version(dist, config)
 

--- a/src/setuptools_scm/_integration/setuptools.py
+++ b/src/setuptools_scm/_integration/setuptools.py
@@ -71,14 +71,17 @@ def version_keyword(
     keyword: str,
     value: bool | dict[str, Any] | Callable[[], dict[str, Any]],
 ) -> None:
+    overrides: dict[str, Any]
     if value is True:
-        value = {}
+        overrides = {}
     elif callable(value):
-        value = value()
+        overrides = value()
+    else:
+        assert isinstance(value, dict), "version_keyword expects a dict or True"
+        overrides = value
 
-    assert isinstance(value, dict), "version_keyword expects a dict or True"
     assert (
-        "dist_name" not in value
+        "dist_name" not in overrides
     ), "dist_name may not be specified in the setup keyword "
     dist_name: str | None = dist.metadata.name
     _log_hookstart("version_keyword", dist)
@@ -86,17 +89,14 @@ def version_keyword(
     if dist.metadata.version is not None:
         warnings.warn(f"version of {dist_name} already set")
         return
-    log.debug(
-        "version keyword %r",
-        vars(dist.metadata),
-    )
-    log.debug("dist %s %s", id(dist), id(dist.metadata))
 
     if dist_name is None:
         dist_name = read_dist_name_from_setup_cfg()
 
     config = _config.Configuration.from_file(
-        dist_name=dist_name, _require_section=False, **value
+        dist_name=dist_name,
+        _require_section=False,
+        **overrides,
     )
     _assign_version(dist, config)
 

--- a/src/setuptools_scm/_integration/toml.py
+++ b/src/setuptools_scm/_integration/toml.py
@@ -6,13 +6,16 @@ from typing import Any
 from typing import Callable
 from typing import cast
 from typing import Dict
+from typing import TYPE_CHECKING
 from typing import TypedDict
 
 if sys.version_info >= (3, 11):
     from tomllib import loads as load_toml
 else:
     from tomli import loads as load_toml
-from typing_extensions import TypeAlias
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
 
 from .. import _log
 

--- a/src/setuptools_scm/_integration/toml.py
+++ b/src/setuptools_scm/_integration/toml.py
@@ -27,7 +27,7 @@ TOML_LOADER: TypeAlias = Callable[[str], TOML_RESULT]
 
 def read_toml_content(path: Path, default: TOML_RESULT | None = None) -> TOML_RESULT:
     try:
-        data = path.read_text()
+        data = path.read_text(encoding="utf-8")
     except FileNotFoundError:
         if default is None:
             raise

--- a/src/setuptools_scm/_integration/toml.py
+++ b/src/setuptools_scm/_integration/toml.py
@@ -1,0 +1,52 @@
+from __future__ import annotations
+
+import sys
+from pathlib import Path
+from typing import Any
+from typing import Callable
+from typing import cast
+from typing import Dict
+from typing import TypedDict
+
+if sys.version_info >= (3, 11):
+    from tomllib import loads as load_toml
+else:
+    from tomli import loads as load_toml
+from typing_extensions import TypeAlias
+
+from .. import _log
+
+log = _log.log.getChild("toml")
+
+TOML_RESULT: TypeAlias = Dict[str, Any]
+TOML_LOADER: TypeAlias = Callable[[str], TOML_RESULT]
+
+
+def read_toml_content(path: Path, default: TOML_RESULT | None = None) -> TOML_RESULT:
+    try:
+        data = path.read_text()
+    except FileNotFoundError:
+        if default is None:
+            raise
+        else:
+            log.debug("%s missing, presuming default %r", path, default)
+            return default
+    else:
+        return load_toml(data)
+
+
+class _CheatTomlData(TypedDict):
+    cheat: dict[str, Any]
+
+
+def load_toml_or_inline_map(data: str | None) -> dict[str, Any]:
+    """
+    load toml data - with a special hack if only a inline map is given
+    """
+    if not data:
+        return {}
+    elif data[0] == "{":
+        data = "cheat=" + data
+        loaded: _CheatTomlData = cast(_CheatTomlData, load_toml(data))
+        return loaded["cheat"]
+    return load_toml(data)

--- a/src/setuptools_scm/_overrides.py
+++ b/src/setuptools_scm/_overrides.py
@@ -3,13 +3,11 @@ from __future__ import annotations
 import os
 import re
 from typing import Any
-from typing import cast
-from typing import TypedDict
 
 from . import _config
 from . import _log
 from . import version
-from ._integration.pyproject_reading import lazy_toml_load
+from ._integration.toml import load_toml_or_inline_map
 
 log = _log.log.getChild("overrides")
 
@@ -49,23 +47,6 @@ def _read_pretended_version_for(
         return version.meta(tag=pretended, preformatted=True, config=config)
     else:
         return None
-
-
-class _CheatTomlData(TypedDict):
-    cheat: dict[str, Any]
-
-
-def load_toml_or_inline_map(data: str | None) -> dict[str, Any]:
-    """
-    load toml data - with a special hack if only a inline map is given
-    """
-    if not data:
-        return {}
-    elif data[0] == "{":
-        data = "cheat=" + data
-        loaded: _CheatTomlData = cast(_CheatTomlData, lazy_toml_load(data))
-        return loaded["cheat"]
-    return lazy_toml_load(data)
 
 
 def read_toml_overrides(dist_name: str | None) -> dict[str, Any]:

--- a/src/setuptools_scm/_version_cls.py
+++ b/src/setuptools_scm/_version_cls.py
@@ -1,6 +1,5 @@
 from __future__ import annotations
 
-from logging import getLogger
 from typing import cast
 from typing import Type
 from typing import Union
@@ -11,6 +10,9 @@ try:
 except ImportError:
     from setuptools.extern.packaging.version import InvalidVersion  # type: ignore
     from setuptools.extern.packaging.version import Version as Version  # type: ignore
+from . import _log
+
+log = _log.log.getChild("version_cls")
 
 
 class NonNormalizedVersion(Version):
@@ -41,10 +43,8 @@ class NonNormalizedVersion(Version):
 def _version_as_tuple(version_str: str) -> tuple[int | str, ...]:
     try:
         parsed_version = Version(version_str)
-    except InvalidVersion:
-        log = getLogger(__name__).parent
-        assert log is not None
-        log.error("failed to parse version %s", version_str)
+    except InvalidVersion as e:
+        log.error("failed to parse version %s: %s", e, version_str)
         return (version_str,)
     else:
         version_fields: tuple[int | str, ...] = parsed_version.release

--- a/testing/conftest.py
+++ b/testing/conftest.py
@@ -20,7 +20,7 @@ def pytest_configure() -> None:
     os.environ["SETUPTOOLS_SCM_DEBUG"] = "1"
 
 
-VERSION_PKGS = ["setuptools", "setuptools_scm", "packaging"]
+VERSION_PKGS = ["setuptools", "setuptools_scm", "packaging", "build", "wheel"]
 
 
 def pytest_report_header() -> list[str]:

--- a/testing/test_integration.py
+++ b/testing/test_integration.py
@@ -112,7 +112,14 @@ def test_pyproject_missing_setup_hook_works(wd: WorkDir, use_scm_version: str) -
     )
     wd.write(
         "pyproject.toml",
-        """[build-system]\nrequires=["setuptools", "setuptools_scm"]\n\n[tool]""",
+        textwrap.dedent(
+            """
+            [build-system]
+            requires=["setuptools", "setuptools_scm"]
+            build-backend = "setuptools.build_meta"
+            [tool]
+            """
+        ),
     )
 
     res = subprocess.run(

--- a/testing/test_integration.py
+++ b/testing/test_integration.py
@@ -128,8 +128,6 @@ def test_pyproject_missing_setup_hook_works(wd: WorkDir, use_scm_version: str) -
     res_build = subprocess.run(
         [sys.executable, "-m", "build", "-nxw"],
         env={k: v for k, v in os.environ.items() if k != "SETUPTOOLS_SCM_DEBUG"},
-        capture_output=True,
-        text=True,
         cwd=wd.cwd,
     )
     import pprint

--- a/testing/test_integration.py
+++ b/testing/test_integration.py
@@ -1,6 +1,8 @@
 from __future__ import annotations
 
 import importlib.metadata
+import os
+import subprocess
 import sys
 import textwrap
 from pathlib import Path
@@ -97,6 +99,44 @@ def test_pyproject_support_with_git(wd: WorkDir, metadata_in: str) -> None:
     wd.write("setup.cfg", SETUP_CFG_FILES[metadata_in])
     res = wd([sys.executable, "setup.py", "--version"])
     assert res.endswith("0.1.dev0+d20090213")
+
+
+@pytest.mark.parametrize("use_scm_version", ["True", "{}", "lambda: {}"])
+def test_pyproject_missing_setup_hook_works(wd: WorkDir, use_scm_version: str) -> None:
+    wd.write(
+        "setup.py",
+        f"""__import__('setuptools').setup(
+    name="example-scm-unique",
+    use_scm_version={use_scm_version},
+    )""",
+    )
+    wd.write(
+        "pyproject.toml",
+        """[build-system]\nrequires=["setuptools", "setuptools_scm"]\n\n[tool]""",
+    )
+
+    res = subprocess.run(
+        [sys.executable, "setup.py", "--version"],
+        cwd=wd.cwd,
+        check=True,
+        stdout=subprocess.PIPE,
+        encoding="utf-8",
+    )
+    stripped = res.stdout.strip()
+    assert stripped.endswith("0.1.dev0+d20090213")
+
+    res_build = subprocess.run(
+        [sys.executable, "-m", "build", "-nxw"],
+        env={k: v for k, v in os.environ.items() if k != "SETUPTOOLS_SCM_DEBUG"},
+        capture_output=True,
+        text=True,
+        cwd=wd.cwd,
+    )
+    import pprint
+
+    pprint.pprint(res_build)
+    wheel: Path = next(wd.cwd.joinpath("dist").iterdir())
+    assert "0.1.dev0+d20090213" in str(wheel)
 
 
 def test_pretend_version(monkeypatch: pytest.MonkeyPatch, wd: WorkDir) -> None:

--- a/tox.ini
+++ b/tox.ini
@@ -13,6 +13,7 @@ deps=
     pytest
     setuptools >= 45
     rich
+    build
 commands=
     pytest {posargs}
 

